tensorflow/compiler/xla/service/algebraic_simplifier.cc

```diff
@@ -80,6 +80,51 @@ bool IsAll(const HloInstruction* op, int8 value) {
   }
 }
 
+bool IsPositive(const HloInstruction* hlo,
+                const AlgebraicSimplifierOptions& options) {
+  switch (hlo->opcode()) {
+    case HloOpcode::kGetTupleElement: {
+      const HloInstruction* gte_operand = hlo->operand(0);
+      switch (gte_operand->opcode()) {
+        case HloOpcode::kCustomCall: {
+          const auto& target = gte_operand->custom_call_target();
+          return target ==
+                     options.get_cudnn_batchnorm_forward_training_metadata() &&
+                 hlo->tuple_index() == 2;
+        }
+        default:
+          return false;
+      }
+    }
+    case HloOpcode::kPower:
+    case HloOpcode::kAbs:
+    case HloOpcode::kRsqrt:
+    case HloOpcode::kSqrt:
+      return IsPositive(hlo->operand(0), options);
+
+    case HloOpcode::kMultiply: {
+      return hlo->operand(0) == hlo->operand(1) &&
+             IsPositive(hlo->operand(0), options);
+    }
+    default:
+      return false;
+  }
+}
+
+bool IsNonNegative(const HloInstruction* hlo,
+                   const AlgebraicSimplifierOptions& options) {
+  switch (hlo->opcode()) {
+    case HloOpcode::kMultiply: {
+      return hlo->operand(0) == hlo->operand(1);
```

Comment:
For kMultiply, we need to make sure that the operand is not a complex number.

Reply:
Infact, in that this case we should handle all cases when the operands are complex i.e, have an assertion that none of the operands can be complex type since `IsPositive` or `IsNonNegative` is not designed to used of Complex dtypes. IMO it would be better to have assertions to not allow complex.

Reply:
Just to clarify, would you prefer an assertion when `IsPositive` or `IsNonNegative` is called with complex args or return false?

Reply:
It can just return false.